[wasm] initial build of coreclr.runtime

radekdoulik opened the pull request.

radekdoulik committed:
First round of changes to build on wasm

radekdoulik committed:
More changes to build on wasm

radekdoulik committed:
Next round of changes

It builds now

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main'

janvorli commented on src/coreclr/clrdefinitions.cmake:
```diff
@@ -119,8 +119,8 @@ if(CLR_CMAKE_TARGET_LINUX)
-if(NOT CLR_CMAKE_TARGET_NETBSD)
-    add_definitions(-DFEATURE_HIJACK)
+if(NOT CLR_CMAKE_TARGET_NETBSD AND NOT CLR_CMAKE_TARGET_ARCH_WASM)
+    add_definitions(-DFEATURE_HIJACK AND NOT CLR_CMAKE_TARGET_ARCH_WASM)
```
This line shouldn't have been changed, it looks like a paste issue

radekdoulik replied:
yeah, last minute change with paste typo

janvorli commented on src/coreclr/gc/env/gcenv.base.h:
```diff
@@ -143,6 +143,8 @@ typedef DWORD (WINAPI *PTHREAD_START_ROUTINE)(void* lpThreadParameter);
   #define MemoryBarrier() { __dmb(_ARM64_BARRIER_SY); }
 
+ #elif defined(HOST_BROWSER)
+  __forceinline void YieldProcessor() { }
```
The runtimelabs NativeAOT llvm has this:
```c++
#if TARGET_WASM
 #define YieldProcessor()
 #define MemoryBarrier __sync_synchronize
#endif // TARGET_WASM
```

janvorli commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -914,7 +914,7 @@ class allocator
             alloc_list* al = &(allocator_to_copy->alloc_list_of (i));
             alloc_list_tail_of(i) = al->alloc_list_tail();
-#if !defined(TARGET_AMD64) && !defined(TARGET_X86)
+#if !defined(TARGET_AMD64) && !defined(TARGET_X86) && !defined(TARGET_BROWSER)
```
If we do the change in gcenv.base.h based on the runtimelabs stuff, this change is not needed.

janvorli commented on src/coreclr/inc/check.inl:
```diff
@@ -316,10 +316,11 @@ inline CHECK CheckUnderflow(const void *address, UINT64 offset)
 {
-#if POINTER_BITS == 32
+    // SIZE_T is 32bit on wasm32
+#if !defined(__wasm__) && POINTER_BITS == 32
```
Why does this need to be disabled? It seems the check should work on wasm too.

radekdoulik replied:
This is to avoid 32bit shift of SIZE_T, which is 32bit on wasm.

janvorli commented on src/coreclr/inc/clrnt.h:
```diff
@@ -501,4 +501,47 @@ RtlVirtualUnwind(
+#define UNW_FLAG_UHANDLER               0x2             /* unwind handler */
+
+PEXCEPTION_ROUTINE
+RtlVirtualUnwind (
```
This should not be needed for WASM. This stuff is windows only.

radekdoulik replied:
This is used in
```
src/coreclr/vm/stackwalk.cpp:516:9: error: use of undeclared identifier 'RtlVirtualUnwind'
    516 |         RtlVirtualUnwind(0,
        |         ^
  1 error generated.
```
and is inside chunk of code with `FEATURE_EH_FUNCLETS` ifdef.

What do you think about this feature for wasm at this point? Should I keep it enabled?

jkotas replied:
Yes, you want to keep `FEATURE_EH_FUNCLETS` enabled.

janvorli commented on src/coreclr/inc/clrtypes.h:
```diff
@@ -338,7 +338,7 @@ inline UINT64 AlignUp(UINT64 value, UINT alignment)
 }
 
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(__wasm__)
```
It is not clear to me why we need this for wasm when it was used on apple only before.

janvorli commented on src/coreclr/inc/corcompile.h:
```diff
@@ -56,6 +56,11 @@ inline ReadyToRunCrossModuleInlineFlags operator &( const ReadyToRunCrossModuleI
 }
 
+#ifdef TARGET_WASM
+// why was it defined only for x86 before?
```
Because all other platforms have it defined in either the pal.h or in the Windows SDK.

radekdoulik replied:
I don't see it in pal.h, so is it in the windows sdk then?

janvorli commented on src/coreclr/inc/daccess.h:
```diff
@@ -1104,7 +1104,7 @@ class __DPtrBase : public __TPtrBase
         return DPtrType(DacTAddrOffset(m_addr, val, sizeof(type)));
     }
-#if defined (HOST_64BIT)
+#if defined (HOST_64BIT) || defined(HOST_WASM)
```
Where is this needed when WASM is a 32 bit platform and no other 32 bit platform needs that?

radekdoulik replied:
Interestingly this doesn't fail anymore. Either it was hit in a code in disabled feature or it changed in main meanwhile. I will remove it.

janvorli commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14811,3 +14811,11 @@ ULONG EECodeInfo::GetFrameOffsetFromUnwindInfo()
 #endif // defined(TARGET_AMD64)
+
+#ifdef TARGET_WASM
+ULONG EECodeInfo::GetFixedStackSize()
```
This should not be needed, as this function is ever called on amd64 and x86.

radekdoulik replied:
I think I was hitting this before I disabled FEATURE_METADATA_UPDATER on wasm.

I removed the change.

janvorli commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -10181,6 +10181,8 @@ void MethodTableBuilder::CheckForSystemTypes()
             // although applying 16 byte alignment is consistent with treatment of 128 bit SSE types
             // even on X86
             pLayout->m_ManagedLargestAlignmentRequirementOfAllMembers = 16; // sizeof(__int128)
+#elif defined(TARGET_WASM)
```
This looks equivalent to the previous #elif, any reason for having it separate?

radekdoulik replied:
Not really, it is just that I would need to update the comment and I am also not sure if it is the final value. I don't know yet if we will need 16 bytes alignment for v128. It looks like 4 might be enough and 8 for int128 which is represented as structure of 2x int64 in wasm.

janvorli commented on src/coreclr/vm/wasm/asmconstants.h:
```diff
@@ -0,0 +1,10 @@
+// Be sure to rebuild clr/src/vm/ceemain.cpp after changing this file, to
+// ensure that the constants match the expected C/C++ values
+
+#define DynamicHelperFrameFlags_ObjectArg   1
```
Where is this needed? These constants are used only by asm helpers.

radekdoulik replied:
I hit it here
```
src/coreclr/vm/frames.cpp:1028:37: error: use of undeclared identifier 'DynamicHelperFrameFlags_ObjectArg'
   1028 |     if (m_dynamicHelperFrameFlags & DynamicHelperFrameFlags_ObjectArg)
```.
Maybe I should disable `FEATURE_READYTORUN`. That would need more changes though IIRC.

janvorli commented on src/coreclr/vm/wasm/cgencpu.h:
```diff
@@ -0,0 +1,214 @@
+    VOID EmitComputedInstantiatingMethodStub(MethodDesc* pSharedMD, struct ShuffleEntry *pShuffleEntryArray, void* extraArg);
+};
+
+struct DECLSPEC_ALIGN(4) UMEntryThunkCode
```
The UMEntryThunkCode was deleted from the main few days ago

jkotas commented on src/native/corehost/hostmisc/pal.h:
```diff
@@ -299,6 +299,7 @@ namespace pal
         s390X,
         x64,
         x86,
+        wasm,
```
We should not need to build native/corehost for wasm.

radekdoulik committed:
Fix debug build

radekdoulik committed:
Fix paste typo

radekdoulik committed:
Feedback

radekdoulik committed:
Feedback

It is not needed anymore

radekdoulik committed:
Feedback

This is not needed anymore as FEATURE_METADATA_UPDATER is now disabled on wasm

radekdoulik committed:
Feedback

UMEntryThunkCode was removed from main recently

radekdoulik committed:
Feedback

Disable FEATURE_VIRTUAL_STUB_DISPATCH and enable FEATURE_CACHED_INTERFACE_DISPATCH

radekdoulik committed:
Feedback

Remove old host changes

radekdoulik committed:
Build runtime on CI

radekdoulik committed:
Inline few functions in header files

radekdoulik committed:
Feedback

Forgot to remove this as well

radekdoulik committed:
Add missing file

radekdoulik committed:
Feedback

Add assert to Thread::VirtualUnwindCallFrame

radekdoulik committed:
Remove unwanted change

radekdoulik committed:
Remove unwanted change

janvorli reviewed (approved):
LGTM, thank you!
